src/mapgen_functions.cpp

```diff
@@ -977,382 +980,319 @@ void mapgen_fungal_flowers(map *m, oter_id, mapgendata dat, int, float)
+                {
+                    int x1=0, y1=0, x2=3, y2=SEEY-1+dead_end_extension;
+                    coord_rotate_cw(x1,y1,dir);
+                    coord_rotate_cw(x2,y2,dir);
+                    square(m, t_sidewalk, x1, y1, x2, y2);
                 }
-            } else {
-                if (((i == SEEX - 1 || i == SEEX) && j % 4 != 0) ||
-                     ((j == SEEY - 1 || j == SEEY) && i % 4 != 0 && i > SEEX)) {
-                    m->ter_set(i, j, t_pavement_y);
-                } else {
-                    m->ter_set(i, j, t_pavement);
+                // sidewalk east of north road, etc
+                if (sidewalks_neswx[(dir+1)%4  ] || // has_sidewalk east?
+                    sidewalks_neswx[(dir+1)%4+4] || // has_sidewalk northeast?
+                    sidewalks_neswx[ dir       ])   // has_sidewalk north?
+                {
+                    int x1=SEEX*2-5, y1=0, x2=SEEX*2-1, y2=SEEY-1+dead_end_extension;
+                    coord_rotate_cw(x1,y1,dir);
+                    coord_rotate_cw(x2,y2,dir);
+                    square(m, t_sidewalk, x1, y1, x2, y2);
                 }
             }
         }
-    }
-    if (terrain_type == "road_esw") {
-        m->rotate(1);
-    }
-    if (terrain_type == "road_nsw") {
-        m->rotate(2);
-    }
-    if (terrain_type == "road_new") {
-        m->rotate(3);
-    }
-    if(sidewalks) {
-        m->place_spawns( mongroup_id( "GROUP_ZOMBIE" ), 2, 0, 0, SEEX * 2 - 1, SEEX * 2 - 1, density);
-        // 1 per 10 overmaps
-        if (one_in(10000)) {
-            m->add_spawn(mon_zombie_jackson, 1, SEEX, SEEY);
-        }
-    }
-    m->place_items("road", 5, 0, 0, SEEX * 2 - 1, SEEX * 2 - 1, false, turn);
-}
 
-void mapgen_road_four_way(map *m, oter_id terrain_type, mapgendata dat, int turn, float density)
-{
-    bool plaza = false;
-    for (int i = 0; i < 4; i++) {
-        if (dat.t_nesw[i] == "road_nesw" || dat.t_nesw[i] == "road_nesw_manhole") {
-            plaza = true;
-        }
-    }
-    bool sidewalks = false;
-    for (int i = 0; i < 8; i++) {
-        if (otermap[dat.t_nesw[i]].has_flag(has_sidewalk)) {
-            sidewalks = true;
+        //draw dead end sidewalk
+        if(dead_end_extension && neighbor_sidewalks)
+            square(m, t_sidewalk, 0, SEEY+dead_end_extension, SEEX*2-1, SEEY+dead_end_extension+4);
+
+        // draw 16-wide pavement from the middle to the edge in each road direction
+        // also corner pieces to curve towards diagonal neighbors
+        for (int dir=0; dir<4; dir++) {
+            if (roads_nesw[dir]) {
+                int x1=4, y1=0, x2=SEEX*2-1-4, y2=SEEY-1+dead_end_extension;
+                coord_rotate_cw(x1,y1,dir);
+                coord_rotate_cw(x2,y2,dir);
+                square(m, t_pavement, x1, y1, x2, y2);
+                if(curvedir_nesw[dir] != 0)
+                    for(int x=1; x<4; x++)
+                        for(int y=0; y<x; y++) {
```

Comment:
also, in `switch (num_dirs)` above, there are a lot of `if(){foo; break;} if(){bar; break;}` that I think make the code easier to read. replacing those with `if() foo; else if() bar;` might not be an improvement.